ovos_utils/bracket_expansion.py

```diff
@@ -0,0 +1,196 @@
+
+    def __repr__(self):
+        return self._tree.__repr__()
+
+
+class Word(Fragment):
+    """
+    Single word in the sentence tree.
+    Construct with a string as argument.
+    """
+
+    def expand(self):
+        """
+        Creates one sentence that contains exactly that word.
+        Returns:
+            List<List<str>>: A list with the given string as sentence
+                                (= token/string list)
+        """
+        return [[self._tree]]
+
+
+class Sentence(Fragment):
+    """
+    A Sentence made of several concatenations/words.
+    Construct with a List<Fragment> as argument.
+    """
+
+    def expand(self):
+        """
+        Creates a combination of all sub-sentences.
+        Returns:
+            List<List<str>>: A list with all subsentence expansions combined in
+                                every possible way
+        """
+        old_expanded = [[]]
+        for sub in self._tree:
+            sub_expanded = sub.expand()
+            new_expanded = []
+            while len(old_expanded) > 0:
+                sentence = old_expanded.pop()
+                for new in sub_expanded:
+                    new_expanded.append(sentence + new)
+            old_expanded = new_expanded
+        return old_expanded
+
+
+class Options(Fragment):
+    """
+    A Combination of possible sub-sentences.
+    Construct with List<Fragment> as argument.
+    """
+
+    def expand(self):
+        """
+        Returns all of its options as seperated sub-sentences.
+        Returns:
+            List<List<str>>: A list containing the sentences created by all
+                                expansions of its sub-sentences
+        """
+        options = []
+        for option in self._tree:
+            options.extend(option.expand())
+        return options
+
+
+class SentenceTreeParser:
+    """
+    Generate sentence token trees from a list of tokens
+    ['1', '(', '2', '|', '3, ')'] -> [['1', '2'], ['1', '3']]
+    """
+
+    def __init__(self, tokens):
+        self.tokens = tokens
+
+    def _parse(self):
+        """
+        Generate sentence token trees
+        ['1', '(', '2', '|', '3, ')'] -> ['1', ['2', '3']]
+        """
+        self._current_position = 0
+        return self._parse_expr()
+
+    def _parse_expr(self):
+        """
+        Generate sentence token trees from the current position to
+        the next closing parentheses / end of the list and return it
+        ['1', '(', '2', '|', '3, ')'] -> ['1', [['2'], ['3']]]
+        ['2', '|', '3'] -> [['2'], ['3']]
+        """
+        # List of all generated sentences
+        sentence_list = []
+        # Currently active sentence
+        cur_sentence = []
+        sentence_list.append(Sentence(cur_sentence))
+        # Determine which form the current expression has
+        while self._current_position < len(self.tokens):
+            cur = self.tokens[self._current_position]
+            self._current_position += 1
```

Comment:
`_current_position` is populated by `_parse()`. Is `_parse()` guaranteed to fire before these lines?